### PR TITLE
Implement torch.tensordot

### LINKED ISSUE #8988: [Feature Request] tensordot
``tensordot`` would be a useful function to have for general contractions between tensors. It looks like all the basic functionality is already present, but not a [numpy-matching api](https://docs.scipy.org/doc/numpy/reference/generated/numpy.tensordot.html). One nice gain if this was present would be that ``pytorch`` could be used as a backend by [``opt_einsum``](https://github.com/dgasmith/opt_einsum) for order-optimized tensor contractions.

Two simple and purely python options would be:

1. a ``dot/matmul`` based method - like [__this gist__](https://gist.github.com/deanmark/9aec75b7dc9fa71c93c4bc85c5438777) (adapted from numpy's implementation itself I think)
2. an ``einsum`` based method - like in [__this PR__](https://github.com/devinamatthews/tblis/pull/17/files) (not for pytorch but similarly translates ``tensordot`` calls to ``einsum`` calls)

### PR DESCRIPTION
**Pull request by t-vi**
Fixes: #8988

### CODE REVIEW

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

From a first look and by comparing side by side against the [numpy implementation](https://github.com/numpy/numpy/blob/a52634d7ea38f74ba286580f96c35190b2b3b549/numpy/core/numeric.py#L1133-L1316) this looks good.

Also, I'd prefer if we named the ATen functions without an underscore.

**Comment by jcmgray** [comment]
Nice! Just wondering if there is a good reason to differ from the numpy api regarding renaming the keyword argument ``axes`` -> ``dims``?

**Comment by t-vi** [comment]
PyTorch consistently uses dim instead of axis, so I'm trying to blend in.

**Comment by t-vi** [comment]
So I don't know that the CI failures are from the patch. Could you have it retested, please?

File: torch/functional.py [hunk 1]
```diff
@@ -459,4 +514,4 @@ def argsort(input, dim=None, descending=False):
     """
     if dim is None:
         return torch.sort(input, -1, descending)[1]
-    return torch.sort(input, dim, descending)[1]
+    return torch.sort(input, dim, descending)[1]
```
Comment (yf225) [inline torch/functional.py]: Lint error: `./torch/functional.py:517:49: W292 no newline at end of file`

File: torch/functional.py [hunk 2]
```diff
@@ -429,6 +430,60 @@ def argmin(input, dim=None, keepdim=False):
...
+      b (Tensor): Right tensor to contract
+      dims (int or tuple of two lists of integers): number of dimensions to
+         contract or explicit lists of dimensions for :attr:`a` and
+         :attr`b` respectively
+
+    When called with an integer argument :attr:`dims` = :math:`d`, and the number of
+    dimensions of :attr:`a` and :attr:`b` is :math:`m` and :math`n`, respectively,
+    it computes
+
+    :math:`r_{i_0,...,i_{m-d}, i_d,...,i_n}`
```
Comment (SsnL) [inline torch/functional.py]: nit: use `.. math::` block

File: torch/functional.py [hunk 3]
```diff
@@ -429,6 +430,60 @@ def argmin(input, dim=None, keepdim=False):
...
+        >>> a = torch.arange(60.).reshape(3, 4, 5)
+        >>> b = torch.arange(24.).reshape(4, 3, 2)
+        >>> torch.tensordot(a, b, dims=([1, 0], [0, 1]))
+        tensor([[4400., 4730.],
+                [4532., 4874.],
+                [4664., 5018.],
+                [4796., 5162.],
+                [4928., 5306.]])
+
+        >>> a = torch.randn(3, 4, 5, device=d)
```
Comment (SsnL) [inline torch/functional.py]: hmm `d`? maybe use `'cuda'`?

File: torch/functional.py [hunk 4]
```diff
@@ -429,6 +430,60 @@ def argmin(input, dim=None, keepdim=False):
...
+
+    When called with an integer argument :attr:`dims` = :math:`d`, and the number of
+    dimensions of :attr:`a` and :attr:`b` is :math:`m` and :math`n`, respectively,
+    it computes
+
+    :math:`r_{i_0,...,i_{m-d}, i_d,...,i_n}`
+    :math:`= \sum_{k_0,...,k_{d-1}} a_{i_0,...,i_{m-d},k_0,...,k_{d-1}} * b_{k_0,...,k_{d-1}, i_d,...,i_n}`.
+
+    When called with :attr:`dims` of the list form, the given dimensions will be contracted
+    in place of the last :math:`d` of :attr:`a` and the first :math:`d` of `b`. The sizes
```
Comment (SsnL) [inline torch/functional.py]: nit: ``:attr:`b` ``

**Review by ssnl — APPROVED** [review]
LGTM other than small nits

**Comment by t-vi** [comment]
Anything I can do to move this further?

**Comment by SsnL** [comment]
@t-vi i'll push this through
